### PR TITLE
[c++] Support `nnz` when `allow_duplicates=True`

### PR DESCRIPTION
**Pull request by johnkerl**
**Issue and/or context:**

#988 

**Changes:**

We've established the pattern that:

* `allow_duplicates=False` is the default
* `allow_duplicates=True` is supported with user caveats that uniqueness is the application's responsibility

Here we honor that pattern.

### CODE REVIEW

File: libtiledbsoma/src/soma_reader.cc [hunk 1]
```diff
@@ -274,32 +274,34 @@ uint64_t SOMAReader::nnz() {
...
-    // If duplicates are allowed, we cannot count simply count cells.
     if (mq_->schema()->allows_dups()) {
-        throw TileDBSOMAError(
-            "[SOMAReader] nnz not supported when duplicates are allowed");
+        LOG_DEBUG(
+            "[SOMAReader] computing nnz() when duplicates are allowed: this is "
+            "accurate only when the application has otherwise ensured "
+            "uniqueness");
     }
+}
```
Comment (gspowley) [inline libtiledbsoma/src/soma_reader.cc:283]: Extra `{`?

**Review by gspowley — APPROVED** [review]
Thanks @johnkerl, for making it clear we are honoring the caveat that uniqueness is the application's responsibility.

**Comment by bkmartinjr** [comment]
@johnkerl - see my comments on #988. I'm still not sure this is right, and would appreciate a chance to work through the consistency issue

**Comment by bkmartinjr** [comment]
Related to my comments in #988, and this code:
```
        // If any relevant fragment is a consolidated fragment, fall back to
        // counting cells, because the fragment may contain duplicates.
        if (frag_ts.first != frag_ts.second) {
            return nnz_slow();
        }
```
I believe that when `allows_duplicates=True`, the call to nnz_slow is going to do the same thing as if we drop through to code below (i.e., do not call `nnz_slow`).  Rationale:  the code in nnz_slow does not de-dup in this corner case.

If you buy my argument in #988, the fix is to not bother calling `nnz_slow` in this case when `allows_duplicates==True`. Suggested modification is something like:
```
        // If any relevant fragment is a consolidated fragment, fall back to
        // counting cells, because the fragment may contain duplicates.
        if (!mq_->schema()->allows_dups() && frag_ts.first != frag_ts.second) {
            return nnz_slow();
        }
```

**Comment by johnkerl** [comment]
> Suggested modification is something like:

@bkmartinjr yes indeed -- @mlin and I have been thinking the same thing -- I'll do this

File: libtiledbsoma/src/soma_reader.cc (hunk 1 above)
Comment (johnkerl) [reply libtiledbsoma/src/soma_reader.cc]: @gspowley yeah I had a bad merge/rebase. Fixed on commit #2.

**Comment by johnkerl** [comment]
@bkmartinjr I'm ready to merge this on @gspowley's approval -- double-checking this works for you -- if not please let me know if you have any suggestions for how to improve the C++ and/or Python unit-test cases to remove any misalignment between the unit-test cases' articulated expectations and yours -- thank you!

**Pull request merged**